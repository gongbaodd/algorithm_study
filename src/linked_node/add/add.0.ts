import { LNode } from "../lib";
import { result } from "./lib";

result((head1, head2) => {
  if (head1.next === null) {
    return head2;
  }

  if (head2.next === null) {
    return head1;
  }

  let c = 0;
  let cursor1: LNode | null = head1.next;
  let cursor2: LNode | null = head2.next;
  const head3 = new LNode(0);
  let cursor3 = head3;

  while (1) {
    const sum = (cursor1?.data ?? 0) + (cursor2?.data ?? 0) + c;
    const node = new LNode(sum % 10);
    c = parseInt((sum / 10).toString(), 10);

    cursor3.next = node;
    cursor3 = node;

    if (cursor1 !== null && cursor2 !== null) {
      cursor1 = cursor1.next;
      cursor2 = cursor2.next;
      continue;
    }

    if (c > 0) {
      continue;
    }

    [cursor1, cursor2]
      .filter((node) => node === null)
      .map((node) => cursor3.next = node?.next ?? null);
    break;
  }

  return head3;
});
