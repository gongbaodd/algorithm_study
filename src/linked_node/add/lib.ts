import { LNode, show, init, info } from "../lib";

const LNODES_1 = [3, 4, 5, 6, 7, 8];
const LNODES_2 = [9, 8, 7, 6, 5];

export function result(fn: (head1: LNode, head2: LNode) => LNode) {
  const [head1, head2] = [init(LNODES_1), init(LNODES_2)];
  info("Head1:");
  show(head1);

  info("Head2:");
  show(head2);

  const head3 = fn(head1, head2);
  info("After:");
  show(head3);
}
