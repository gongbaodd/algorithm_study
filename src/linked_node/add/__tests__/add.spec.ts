import { add } from "../add.0";
import { init } from "../../lib";

const LNODES_1 = [3, 4, 5, 6, 7, 8];
const LNODES_2 = [9, 8, 7, 6, 5];

describe("add tow linked lists", () => {
  it("should return [2, 3, 3, 3, 3, 9]", () => {
    const head = add(init(LNODES_1), init(LNODES_2));
    expect(head?.next?.data).toEqual(2);
    expect(head).toEqual(init([2, 3, 3, 3, 3, 9]));
    expect(head).toMatchSnapshot();
  });
});
