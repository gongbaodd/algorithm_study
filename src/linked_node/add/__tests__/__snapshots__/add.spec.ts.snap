// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`add tow linked lists should return [2, 3, 3, 3, 3, 9] 1`] = `
LNode {
  "data": 0,
  "next": LNode {
    "data": 2,
    "next": LNode {
      "data": 3,
      "next": LNode {
        "data": 3,
        "next": LNode {
          "data": 3,
          "next": LNode {
            "data": 3,
            "next": LNode {
              "data": 9,
              "next": null,
            },
          },
        },
      },
    },
  },
}
`;
